app/actions/remote/file.ts

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {ClientResponse, ClientResponseError} from '@mattermost/react-native-network-client';
+
+import {Client} from '@app/client/rest';
+import ClientError from '@app/client/rest/error';
+import NetworkManager from '@init/network_manager';
+
+export const uploadFile = (
+    serverUrl: string,
+    file: FileInfo,
+    channelId: string,
+    onProgress: (fractionCompleted: number) => void,
+    onComplete: (response: ClientResponse) => void,
+    onError: (response: ClientResponseError) => void,
+) => {
+    let client: Client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error: error as ClientError};
+    }
+    return {cancel: client.uploadFile(file, channelId, onProgress, onComplete, onError)};
```

Comment:
why don't you return the promise instead and have the callbacks implemented on each component? I mean this works for me, but wanted to understand the reasoning

Reply:
Not sure I understood your question, since there are several levels here.

First of all, I am following the pattern of `return {something?: SomeType, error?: any}`. That way, makes it simpler check whether is an error in the caller.

Apart of that, my original idea was to "reduce complexity on the caller". Right now the only complexity we are removing is the try catch block for getting the client. I still like it better this way, but I can see it is not giving tons of value right now.